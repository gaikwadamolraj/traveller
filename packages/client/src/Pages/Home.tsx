import { useReducer, useRef, useState } from 'react'
import type { Reducer } from 'react'
import type { FC } from 'react'
import { Container, InputRightElement, Input, Heading, InputGroup, IconButton, VStack } from '@chakra-ui/react'
import { Search2Icon } from '@chakra-ui/icons'
import { TablePagination } from '../components/TablePagination'
import type { IAction, ICity, PutPayload, IState } from '../dataStructure'
import { CONSTANTS } from '../dataStructure'
import { sendApiCall } from '../httputil/Api'
import { Loading } from '../components/Loading'
import { Error } from '../components/Error'
import { initialState, reducer, setError } from '../store'

export const Home: FC = () => {
  const [pageNumber, setPage] = useState<number>(CONSTANTS.DEFAULT_PAGE)
  const [serachName, setSearchName] = useState<string>('')

  // eslint-disable-next-line
  const serachInput = useRef<any>()

  const [state, dispatch] = useReducer<Reducer<IState, IAction>>(reducer, initialState)
  const { cities, isLoading, error } = state

  const fetchData = () => {
    const name: string = serachInput?.current?.value.trim()
    setSearchName(name)
    if (name) {
      sendApiCall({ dispatch, urlData: `?name=${name}` })
      setPage(CONSTANTS.DEFAULT_PAGE)
    } else {
      setError(dispatch, 'Please enter some char values')
    }
  }
  const updateData = async (city: ICity, page: number, payload: PutPayload) => {
    sendApiCall({ method: 'PUT', dispatch, urlData: `/${city.id}`, payload })
    setPage(page)
  }

  const handleVisited = (city: ICity, page: number) => {
    updateData(city, page, { visited: !city.visited })
  }

  const handleWishlist = (city: ICity, page: number) => {
    updateData(city, page, { wishlist: !city.wishlist })
  }

  // useEffect(() => {
  //   // fetchData()
  //   return () => {
  //     // Clean up the subscription if any
  //   }
  // }, [])

  return (
    <VStack>
      <Heading as="h1">Smart traveller</Heading>
      <Container maxW="container.md">
        <InputGroup>
          <Input
            placeholder={'Find city by name'}
            ref={serachInput}
            onKeyPress={e => (e.key === 'Enter' ? fetchData() : '')}
          />
          <InputRightElement children={<IconButton aria-label="" icon={<Search2Icon />} />} onClick={fetchData} />
        </InputGroup>
        <Error error={error} />

        {isLoading ? (
          <Loading />
        ) : serachName ? (
          <TablePagination
            pageNumber={pageNumber}
            cities={cities || []}
            handleVisited={handleVisited}
            handleWishlist={handleWishlist}
          />
        ) : (
          ''
        )}
      </Container>
    </VStack>
  )
}
