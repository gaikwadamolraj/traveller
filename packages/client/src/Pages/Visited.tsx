import { useEffect, useReducer } from 'react'
import type { Reducer } from 'react'
import { Container, Heading } from '@chakra-ui/react'
import type { IAction, IEmptyProps, IState } from '../dataStructure'
import { TablePagination } from '../components/TablePagination'
import { Loading } from '../components/Loading'
import { sendApiCall } from '../httputil/Api'
import { Error } from '../components/Error'
import { initialState, reducer } from '../store'

export const Visited: React.FC<IEmptyProps> = () => {
  const [state, dispatch] = useReducer<Reducer<IState, IAction>>(reducer, initialState)
  const { cities, isLoading, error } = state

  useEffect(() => {
    sendApiCall({ dispatch, urlData: `/?visited=true` })
  }, [])

  return (
    <>
      <Heading as="h1">Visited</Heading>
      <Container maxW="container.md">
        <Error error={error} />
        {!isLoading ? <TablePagination cities={cities || []} isLoading /> : <Loading />}
      </Container>
    </>
  )
}
