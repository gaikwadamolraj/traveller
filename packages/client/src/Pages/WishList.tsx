import { useEffect, useReducer } from 'react'
import type { Reducer } from 'react'
import { Container, Heading } from '@chakra-ui/react'
import { TablePagination } from '../components/TablePagination'
import type { IAction, IEmptyProps, IState } from '../dataStructure'
import { sendApiCall } from '../httputil/Api'
import { Loading } from '../components/Loading'
import { Error } from '../components/Error'
import { initialState, reducer } from '../store'

export const WishList: React.FC<IEmptyProps> = () => {
  const [state, dispatch] = useReducer<Reducer<IState, IAction>>(reducer, initialState)

  const { cities, isLoading, error } = state
  useEffect(() => {
    sendApiCall({ dispatch, urlData: `/?wishlist=true` })
  }, [])

  return (
    <>
      <Heading as="h1">Wish list</Heading>
      <Container maxW="container.md">
        <Error error={error} />

        {!isLoading ? <TablePagination cities={cities || []} isLoading /> : <Loading />}
      </Container>
    </>
  )
}
