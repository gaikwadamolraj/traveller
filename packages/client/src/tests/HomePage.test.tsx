import { Home } from '../Pages/Home'
import { render } from '../test-utils'

describe('Home Page component', () => {
  it('Home page', () => {
    const { container } = render(<Home />)
    expect(container).toMatchSnapshot()
  })
})
