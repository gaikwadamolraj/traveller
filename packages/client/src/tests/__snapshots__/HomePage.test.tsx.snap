// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page component Home page 1`] = `
<div>
  <div
    class="chakra-stack css-owjkmg"
  >
    <h1
      class="chakra-heading css-1dklj6k"
    >
      Smart traveller
    </h1>
    <div
      class="chakra-container css-nm5t63"
    >
      <div
        class="chakra-input__group css-4302v8"
      >
        <input
          class="chakra-input css-18g31kk"
          placeholder="Find city by name"
        />
        <div
          class="chakra-input__right-element css-ur97uj"
        >
          <button
            aria-label=""
            class="chakra-button css-dn5esd"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon css-onkibi"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M23.414,20.591l-4.645-4.645a10.256,10.256,0,1,0-2.828,2.829l4.645,4.644a2.025,2.025,0,0,0,2.828,0A2,2,0,0,0,23.414,20.591ZM10.25,3.005A7.25,7.25,0,1,1,3,10.255,7.258,7.258,0,0,1,10.25,3.005Z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
      
    </div>
  </div>
</div>
`;
